File: IKeep.Web/ClientApp/src/models/GenericChore.js
import Entity from './Entity'

class GenericChore extends Entity
{
    get FormatDescription()
    {
        return this.Title + ", " + this.Author;
    }
    constructor(json)
    {
        super(json);
        if(json)
        {
            this.Ref = json.ref;
            this.Description = json.description;
            this.Duration = json.duration;
            this.Period = json.period;
            this.Priority = json.priority;
            this.ChoreTypeId = json.choreTypeId;
            this.FormatLabelId = json.formatId;
            this.CategoryId = json.categoryId;
            this.ChoresIds = json.tasksIds;

            this.ChoreTypeName = json.choreTypeName;
            this.FormatName = json.formatName;
            this.CategoryName = json.categoryName;

            this.GenericChoreFormatLabels = json.genericChoreFormatLabels;
        }
        else
        {
            this.Ref = "";
            this.Description = "";
            this.Duration = "";
            this.Period = Period;
            this.Priority = Priority;
            this.ChoreTypeId = "";
            this.FormatId = "";
            this.CategoryId = "";
            this.ChoresIds = "";

            this.ChoreTypeName = "";
            this.FormatName = "";
            this.CategoryName = "";
            
            this.GenericChoreFormatLabels = [];
        }
    }

}
export default GenericChore;

const Period = Object.freeze({
    Daily: 0,
    Weekly: 1,
    Monthly: 2,
    Bimonthly: 3,
    Quarterly: 4,
    Semester: 5,
    Yearly: 6,
    TwoYearly: 7,
    FourYearly: 8,
    DEFAUL: 9
})

const Priority = Object.freeze({
    Top: 0,
    Average: 1,
    Low: 2,
    DEFAULT: 3
})

File: IKeep.Lib/Models/GenericChore.cs
using IKeep.Lib.Core;
using IKeep.Lib.Models;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace IKeep.Lib.Models
{
    public class GenericChore : Entity
    {
        public string Ref { get; set; }
        public string Description { get; set; }
        public TimeSpan Duration { get; set; }
        public Period Period { get; set; }
        public Priority Priority { get; set; }
        public Guid ChoreTypeId { get; set; }
        [JsonIgnore]
        public virtual ChoreType ChoreType { get; set; }
        public Guid? SupplierId { get; set; }
        [JsonIgnore]
        public virtual Supplier Supplier { get; set; }
        
        public string ChoreTypeName
        {
            get
            {
                return ChoreType == null ? "" : ChoreType.Name;
            }
        }

        public virtual ICollection<GenericChoreFormatLabel> GenericChoreFormatLabels { get; set; }

        public Guid CategoryId { get; set; }
        [JsonIgnore]
        public virtual Category Category { get; set; }

        public string CategoryName
        {
            get
            {
                return Category == null ? "" : Category.Name;
            }
        }

        [JsonIgnore]
        public virtual ICollection<Chore> Chores { get; set; }
        

        [JsonIgnore]
        public virtual ICollection<GenericElementGenericChore> GenericElementGenericChores { get; set; }
        public List<Guid> GenericElementGenericChoresIds
        {
            get
            {
                return GenericElementGenericChores == null ? new List<Guid>() : GenericElementGenericChores.Select(x => x.Id).ToList();
            }
        }

        [JsonIgnore]
        public virtual ICollection<ElementGenericChore> ElementGenericChores { get; set; }
    }

    public enum Period
    {
        Daily,
        Weekly,
        Monthly,
        Bimonthly,
        Quarterly,
        Semester,
        Yearly,
        TwoYearly,
        FourYearly,
        DEFAULT
    }

    public enum Priority
    {
        Top,
        Average,
        Low,
        DEFAULT
    }

}
